packages/expo-intent-launcher/android/src/main/java/expo/modules/intentlauncher/IntentLauncherModule.kt

```diff
@@ -0,0 +1,134 @@
+) : ExportedModule(context), ActivityEventListener {
+  private var pendingPromise: Promise? = null
+  private val uiManager: UIManager? by moduleRegistry()
+  private val activityProvider: ActivityProvider? by moduleRegistry()
+
+  private inline fun <reified T> moduleRegistry() = moduleRegistryDelegate.getFromModuleRegistry<T>()
+
+  override fun getName() = NAME
+
+  override fun onCreate(moduleRegistry: ModuleRegistry) {
+    moduleRegistryDelegate.onCreate(moduleRegistry)
+  }
+
+  @ExpoMethod
+  fun startActivity(activityAction: String, params: ReadableArguments, promise: Promise) {
+    if (pendingPromise != null) {
+      promise.reject(ActivityAlreadyStartedException())
+      return
+    }
+
+    val activity = activityProvider?.currentActivity
+    if (activity == null) {
+      promise.reject(CurrentActivityNotFoundException())
+      return
+    }
+
+    val uiManager = this.uiManager
+    if (uiManager == null) {
+      promise.reject(ModuleNotFoundException("UIManager"))
+      return
+    }
+
+    val intent = Intent(activityAction)
+    if (params.containsKey(ATTR_CLASS_NAME)) {
+      val componentName =
+        if (params.containsKey(ATTR_PACKAGE_NAME)) ComponentName(params.getString(ATTR_PACKAGE_NAME), params.getString(ATTR_CLASS_NAME))
+        else ComponentName(context, params.getString(ATTR_CLASS_NAME))
+      intent.component = componentName
+    }
+
+    // `setData` and `setType` are exclusive, so we need to use `setDateAndType` in that case.
+    if (params.containsKey(ATTR_DATA) && params.containsKey(ATTR_TYPE)) {
+      intent.setDataAndType(Uri.parse(params.getString(ATTR_DATA)), params.getString(ATTR_TYPE))
+    } else {
+      if (params.containsKey(ATTR_DATA)) {
+        intent.data = Uri.parse(params.getString(ATTR_DATA))
+      } else if (params.containsKey(ATTR_TYPE)) {
+        intent.type = params.getString(ATTR_TYPE)
+      }
+    }
+    if (params.containsKey(ATTR_EXTRA)) {
+      intent.putExtras(params.getArguments(ATTR_EXTRA).toBundle())
+    }
+    if (params.containsKey(ATTR_FLAGS)) {
+      intent.addFlags(params.getInt(ATTR_FLAGS))
+    }
+    if (params.containsKey(ATTR_CATEGORY)) {
+      intent.addCategory(params.getString(ATTR_CATEGORY))
+    }
+
+    uiManager.registerActivityEventListener(this)
+    pendingPromise = promise
+
+    try {
+      activity.startActivityForResult(intent, REQUEST_CODE)
+    } catch (e: ActivityNotFoundException) {
+      promise.reject(e)
+      pendingPromise = null
+    }
+  }
+
+  //region ActivityEventListener
+
+  override fun onActivityResult(activity: Activity, requestCode: Int, resultCode: Int, intent: Intent?) {
+    if (requestCode != REQUEST_CODE) return
+
+    val response = Bundle().apply {
+      putInt("resultCode", resultCode)
+      if (intent != null) {
+        val data = intent.data
+        val extras = intent.extras
+        if (data != null) putString(ATTR_DATA, data.toString())
+        if (extras != null) putBundle(ATTR_EXTRA, extras)
```

Comment:
```suggestion
        intent.data?.let {
          putString(ATTR_DATA, it.toString())
        }
        intent.extras?.let {
          putBundle(ATTR_EXTRA, it)
        }
```
You can also inline the let block if you want 😉